os/fs/smartfs/smartfs_smart.c

```diff
@@ -319,39 +319,34 @@ static int smartfs_open(FAR struct file *filep, const char *relpath, int oflags,
 	sf->currsector = sf->entry.firstsector;
 	sf->byteswritten = 0;
 
+	/* Test if the file is opened in APPEND mode. If yes, then seek to the end of the file.
+	 * Seeking will also automatically update the header in the sector buffer
+	 * */
+	if (oflags & O_APPEND) {
+		smartfs_seek_internal(fs, sf, 0, SEEK_END);
+	}
+
 #ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
-	if ((sf->bflags & SMARTFS_BFLAG_DIRTY) == 0) {
-		/* When using sector buffering, current sector with its header should
-		 * always be present in sf->buffer. Otherwise data corruption may arise
-		 * when writing.
+	/* When using sector buffering, current sector with its header should
+	 * always be present in sf->buffer. Otherwise data corruption may arise
+	 * when writing.
+	 */
+	else if ((sf->bflags & SMARTFS_BFLAG_DIRTY) == 0 && (sf->currsector != SMARTFS_ERASEDSTATE_16BIT)) {
+		/* If the sector buffer flags are dirty, then the logic reponsible for writing to
+		 * the buffer and dirtying the flags is expected to have updated the header.
 		 */
-
-		if (sf->currsector != SMARTFS_ERASEDSTATE_16BIT) {
+		if ((oflags & (O_CREAT | O_TRUNC)) == 0) {
 			/* For truncate, we already read header and set data in sf->buffer */
-			if (((oflags & (O_CREAT | O_TRUNC)) == 0) || (oflags & O_APPEND)) {
```

Comment:
we always refer to nuttx's changes, so I think you can just remove offlags & O_APPEND here to avoid many changes..

Reply:
Updated in a different way